### ISSUE TITLE
Dynamically refresh configuration secrets in the `garden` namespace

### ISSUE THREAD

**Issue by rfranzke** [gardener/gardener#1725]
Today, the gardenlet does read secrets for default domains, internal domain, alerting, etc. only once during start-up, and then never again.
This forces administrators to manually restart the gardenlet in all seed clusters in case he/she changes such a secret.
Although this scenario happens very rarely, we should enable the gardenlet to dynamically refresh the secrets such that a restart is not required.

**Comment by rfranzke** [comment]
/touch
We are considering moving these secrets into a separate namespace to prevent allowing the gardenlet accessing the `garden` namespace entirely as it also might contain other sensitive credentials that it might not even need to do it's work (e.g., other kubeconfigs of shooted seeds, etc.), see also #1723

**Comment by timebertt** [comment]
Hmm, why should we move those secrets out of the `garden` namespace?

> as it also might contain other sensitive credentials that it might not even need to do it's work (e.g., other kubeconfigs of shooted seeds, etc.)

Isn't this exactly what the `SeedAuthorizer` (#1723) should check and prevent?

**Comment by vlerenc** [comment]
@timebertt That's only about mutating operations, but you cannot put read operations under lock. The goal is to prevent that an attacker can reach from one seed to other seeds or shared information resp. that shared information exists in the first place. If all gardenlets can read the garden namespace in the garden cluster, that would violate the goal above.

**Comment by timebertt** [comment]
Hmm, the `NodeAuthorizer` also seems to authorize read operations, see [here](https://kubernetes.io/docs/reference/access-authn-authz/node/) and [here](https://github.com/kubernetes/kubernetes/blob/9d99dbc357ed85bcc963fc4bab8a4a3089c910b2/plugin/pkg/auth/authorizer/node/node_authorizer.go#L175).

**Comment by rfranzke** [comment]
I just noted it down because we discussed it, but we will have to revisit this anyways when we get to #1723. Actually, then name `SeedAuthorizer` suggests that it only handles "seed-related" resources, but these general secrets are "global" configuration. Anyways, just notes so that we don't forget it, the details need to be clarified as part of #1723.

**Comment by vlerenc** [comment]
@timebertt https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/
> Admission controllers limit requests to create, delete, modify or connect to (proxy). They do not support read requests.

---

> Hmm, the NodeAuthorizer also seems to authorize read operations, see here and here.
But that's a special hack in Kubernetes that is in from the early days.

Don't get me wrong, if there is a means to limit read requests to only resources a seed/`gardenlet` has a need-to-know and nothing else, great. Can this be done/how?

**Comment by timebertt** [comment]
> @timebertt https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/
> 
> > Admission controllers limit requests to create, delete, modify or connect to (proxy). They do not support read requests.

Ok, I see the title of #1723 is confusing. We don't need a new admission plugin, but rather a new authorizer.
Admission is too late in the request chain and such decisions have to be done in the Authorization layer.

> if there is a means to limit read requests to only resources a seed/`gardenlet` has a need-to-know and nothing else, great. Can this be done/how?

I've done some quick scan through the API server code.
The extended API server (g-api) delegates Authorization decisions to the kube-apiserver by default.
So this should be possible by adding a autorization webhook to the garden cluster: https://kubernetes.io/docs/reference/access-authn-authz/webhook/
Another way might be a bit hacky but still possible to add a new in-tree authorization plugin when constructing g-api's authorization config here: https://github.com/gardener/gardener/blob/fe441a3dac2c5acd68224d5d95f9c0eab541d4ee/cmd/gardener-apiserver/app/gardener_apiserver.go#L316

TL;DR: it should be possible and we should rather restrict access to the resources mentioned in this issue via the `SeedAuthorizer` and not move secrets into other namespaces or similar...

**Comment by vlerenc** [comment]
@timebertt, great, thank you!

**Comment by timuthy** [comment]
After gaining more insights with @rfranzke, we're now confident that we indeed need to copy the secrets into a dedicated namespace as a list/watch operation for secrets cannot be restricted or checked by a custom authorization plugin.

We decided to have a controller for the Gardener Controller manager that watches `Seed` objects.
Responsibilities:
- create a namespace in the Garden cluster `seed-<seed-name>` with `ownerReference` to seed object
- copy secrets from [ReadGardenSecrets](https://github.com/gardener/gardener/blob/5600eb6454d0c94d9c0bf8ba749d35728590db41/pkg/operation/garden/garden.go#L168) e.g., internal domain, default domain, altering secrets from the `garden` namespace to the seed's namespace

A controller in the Gardenlet will watch `Secrets` in the seed's namespace and update an internal map of these secrets that are used by further Gardenlet controllers.

/assign
